File: lock.nim
import winim/lean
import strformat
import strutils
import json
import parseopt


type
  OSVERSIONINFOEX* = object
    dwOSVersionInfoSize*: DWORD
    dwMajorVersion*: DWORD
    dwMinorVersion*: DWORD
    dwBuildNumber*: DWORD
    dwPlatformId*: DWORD
    szCSDVersion*: array[128, WCHAR]
    wServicePackMajor*: WORD
    wServicePackMinor*: WORD
    wSuiteMask*: WORD
    wProductType*: BYTE
    wReserved*: BYTE


type MEMORY_INFORMATION_CLASS* = enum
    MemoryBasicInformation = 0


proc RtlGetVersion*(lpVersionInformation: var OSVERSIONINFOEX): NTSTATUS {.discardable, dynlib: "ntdll", importc: "RtlGetVersion".}
proc NtReadVirtualMemory(ProcessHandle: HANDLE, BaseAddress: PVOID, Buffer: PVOID, BufferSize: SIZE_T, NumberOfBytesRead: PSIZE_T): NTSTATUS {.discardable, dynlib: "ntdll", importc: "NtReadVirtualMemory".}
proc NtQueryInformationProcess(ProcessHandle: HANDLE, ProcessInformationClass: PROCESSINFOCLASS, ProcessInformation: PVOID, ProcessInformationLength: ULONG, ReturnLength: PULONG): NTSTATUS {.discardable, dynlib: "ntdll", importc: "NtQueryInformationProcess".}
proc NtRemoveProcessDebug*(ProcessHandle: HANDLE, DebugObjectHandle: HANDLE): NTSTATUS {.discardable, dynlib: "ntdll", importc: "NtRemoveProcessDebug".}
proc NtTerminateProcess*(ProcessHandle: HANDLE, ExitStatus: NTSTATUS): NTSTATUS {.discardable, dynlib: "ntdll", importc: "NtTerminateProcess".}
proc NtWriteVirtualMemory*(ProcessHandle: HANDLE, BaseAddress: PVOID, Buffer: PVOID, BufferSize: SIZE_T, NumberOfBytesWritten: PSIZE_T): NTSTATUS {.discardable, dynlib: "ntdll", importc: "NtWriteVirtualMemory".}
proc NtProtectVirtualMemory*(ProcessHandle: HANDLE, BaseAddress: ptr PVOID, RegionSize: ptr SIZE_T, NewProtect: ULONG, OldProtect: PULONG): NTSTATUS {.discardable, dynlib: "ntdll", importc: "NtProtectVirtualMemory".}


proc getBuildNumber*(): OSVERSIONINFOEX =
  var osVersionInfo: OSVERSIONINFOEX
  osVersionInfo.dwOSVersionInfoSize = DWORD sizeof(OSVERSIONINFOEX)
  discard RtlGetVersion(osVersionInfo)
  result = osVersionInfo


proc writeToFile*(path: string, content: string) =
  writeFile(path, content)
  echo "[+] File ", path, " generated correctly"


proc lock*(fileName: string) =
  let osVersionInfo = getBuildNumber()
  let versionData = %*{
    "field0": $osVersionInfo.dwMajorVersion,
    "field1": $osVersionInfo.dwMinorVersion,
    "field2": $osVersionInfo.dwBuildNumber,
  }  
  let wrapper = %*[versionData]
  writeToFile(fileName, $wrapper)


proc readRemoteIntPtr*(hProcess: HANDLE, memAddress: PVOID): PVOID =
    var
        buff: array[8, BYTE]
        bytesRead: SIZE_T
    
    let ntstatus = NtReadVirtualMemory(
        hProcess,
        memAddress,
        addr buff[0],
        sizeof(buff).SIZE_T,
        addr bytesRead
    )

    if ntstatus != 0 and ntstatus != 0xC0000005 and ntstatus != 0x8000000D and hProcess != 0:
        echo fmt"[-] Error calling NtReadVirtualMemory (readRemoteIntPtr). NTSTATUS: 0x{cast[int](ntstatus):X} reading address 0x{cast[int](memAddress):X}"
        return nil
    
    result = cast[PVOID](cast[ptr int64](addr buff[0])[])


proc readRemoteWStr*(hProcess: HANDLE, memAddress: PVOID): string =
    var
        buff: array[256, BYTE]
        bytesRead: SIZE_T
    
    let ntstatus = NtReadVirtualMemory(
        hProcess,
        memAddress,
        addr buff[0],
        sizeof(buff).SIZE_T,
        addr bytesRead
    )

    if ntstatus != 0 and ntstatus != 0xC0000005 and ntstatus != 0x8000000D and hProcess != 0:
        echo fmt"[-] Error calling NtReadVirtualMemory (readRemoteWStr). NTSTATUS: 0x{cast[int](ntstatus):X} reading address 0x{cast[int](memAddress):X}"

    var unicodeStr = ""
    var i = 0
    while i < sizeof(buff) - 1:
        if buff[i] == 0 and buff[i+1] == 0:
            break
        
        let wch = cast[ptr WCHAR](addr buff[i])[]
        unicodeStr.add(char(wch))
        i += 2
    
    return unicodeStr


proc custom_get_module_address*(h_process: HANDLE, module_name: string): uint64 =
    const
        process_basic_information_size = 48'u32
        peb_offset = 0x8
        ldr_offset = 0x18
        in_initialization_order_module_list_offset = 0x30
        flink_dllbase_offset = 0x20
        flink_buffer_offset = 0x50
    
    var
        pbi_byte_array: array[process_basic_information_size, BYTE]
        return_length: ULONG = 0
        ntstatus: NTSTATUS
    
    # Query process information
    ntstatus = NtQueryInformationProcess(
        h_process,
        0.PROCESSINFOCLASS,
        cast[PVOID](addr pbiByteArray[0]), #addr pbi_byte_array[0].PVOID,
        process_basic_information_size.ULONG,
        addr returnLength # addr return_length.PULONG
    )
    
    if ntstatus != 0:
        echo "[-] Error calling NtQueryInformationProcess. NTSTATUS: 0x", toHex(cast[int](ntstatus))
        return 0'u64
    
    let peb_pointer = cast[PVOID](cast[uint64](addr pbi_byte_array[0]) + peb_offset)
    let currentProcess = cast[HANDLE](-1)  # 0xffffffff... = -1
    
    # Read PEB address
    var peb_address = cast[uint64](readRemoteIntPtr(currentProcess, peb_pointer))
    if peb_address == 0:
        return 0'u64
    
    # Read LDR address
    let ldr_pointer = cast[PVOID](peb_address + ldr_offset)
    var ldr_address = cast[uint64](readRemoteIntPtr(h_process, ldr_pointer))
    if ldr_address == 0:
        return 0'u64
    
    # Get module list
    let in_initialization_order_module_list = ldr_address + in_initialization_order_module_list_offset
    var next_flink = cast[uint64](readRemoteIntPtr(h_process, cast[PVOID](in_initialization_order_module_list)))
    var dll_base = high(uint64)
    
    while dll_base != 0:
        next_flink -= 0x10
        
        # Read DLL base name
        let buffer = cast[uint64](readRemoteIntPtr(h_process, cast[PVOID](next_flink + flink_buffer_offset)))
        var base_dll_name = ""
        
        if buffer != 0:
            base_dll_name = readRemoteWStr(h_process, cast[PVOID](buffer))
        
        if base_dll_name == module_name:
            # Get DLL base address
            dll_base = cast[uint64](readRemoteIntPtr(h_process, cast[PVOID](next_flink + flink_dllbase_offset)))
            return dll_base
        
        next_flink = cast[uint64](readRemoteIntPtr(h_process, cast[PVOID](next_flink + 0x10)))
    
    return 0'u64



proc get_text_section_info*(ntdll_address: PVOID): array[2, uint32] =
    let h_process = cast[HANDLE](-1)
    var e_lfanew_data: array[4, BYTE]
    let e_lfanew_address = cast[PVOID](cast[uint64](ntdll_address) + 0x3C)
    
    var bytesRead: SIZE_T
    discard NtReadVirtualMemory(
        h_process,
        e_lfanew_address,
        addr e_lfanew_data[0],
        4,
        addr bytesRead
    )
    
    let e_lfanew = cast[ptr uint32](addr e_lfanew_data[0])[]
    
    # Calculate NT headers address
    let nt_headers_address = cast[uint64](ntdll_address) + e_lfanew
    let optional_header_address = nt_headers_address + 24
    
    # Read SizeOfCode
    let sizeofcode_address = cast[PVOID](optional_header_address + 4)
    var sizeofcode_data: array[4, BYTE]
    
    discard NtReadVirtualMemory(
        h_process,
        sizeofcode_address,
        addr sizeofcode_data[0],
        sizeof(sizeofcode_data),
        addr bytesRead
    )
    
    let sizeofcode = cast[ptr uint32](addr sizeofcode_data[0])[]
    
    # Read BaseOfCode
    let baseofcode_address = cast[PVOID](optional_header_address + 20)
    var baseofcode_data: array[4, BYTE]
    
    discard NtReadVirtualMemory(
        h_process,
        baseofcode_address,
        addr baseofcode_data[0],
        sizeof(baseofcode_data),
        addr bytesRead
    )
    
    let baseofcode = cast[ptr uint32](addr baseofcode_data[0])[]
    
    # Return the two values as an array
    [baseofcode, sizeofcode]


proc get_ntdll_from_debug_proc*(process_path: string): ptr uint8 =
    var
        si: STARTUPINFOW
        pi: PROCESS_INFORMATION
        debug_object_handle: HANDLE
        return_length: ULONG
        status: NTSTATUS

    # Initialize structures
    zeroMem(addr si, sizeof(si).SIZE_T)
    si.cb = sizeof(si).DWORD
    zeroMem(addr pi, sizeof(pi).SIZE_T)

    # Create process with DEBUG_PROCESS flag
    let success = CreateProcessW(
        newWideCString(process_path),
        nil,
        nil,
        nil,
        false,
        DEBUG_PROCESS,
        nil,
        nil,
        addr si,
        addr pi
    )
    if success == 0:
        let err = GetLastError()
        echo "[-] CreateProcess failed with error: 0x", toHex(err.uint32), " (", err, ")"
        quit(1)

    # Get ntdll.dll information from current process
    let current_process = cast[HANDLE](-1)
    let local_ntdll_handle = custom_get_module_address(current_process, "ntdll.dll")
    if local_ntdll_handle == 0:
        echo "[-] Failed to locate ntdll.dll in current process"
        quit(1)
    let text_section = get_text_section_info(cast[PVOID](local_ntdll_handle))
    let local_ntdll_txt_base = text_section[0]
    let local_ntdll_txt_size = text_section[1]
    let local_ntdll_txt = local_ntdll_handle + local_ntdll_txt_base

    # Read remote ntdll text section
    var ntdll_buffer = newSeq[byte](local_ntdll_txt_size)
    var bytesRead: SIZE_T
    status = NtReadVirtualMemory(
        pi.hProcess,
        cast[PVOID](local_ntdll_txt),
        addr ntdll_buffer[0],
        local_ntdll_txt_size.SIZE_T,
        addr bytesRead
    )
    if status != 0:
        echo "[-] Read operation failed"
        quit(1)

    # Get debug object handle
    status = NtQueryInformationProcess(
        pi.hProcess,
        30.PROCESSINFOCLASS,
        addr debug_object_handle,
        sizeof(debug_object_handle).ULONG,
        addr return_length
    )
    if status != 0:
        echo "[-] Failed to get debug object handle"
        quit(1)

    # Cleanup and terminate debug process
    status = NtRemoveProcessDebug(pi.hProcess, debug_object_handle)
    if status != 0:
        echo "[-] Failed to remove process debug"
        quit(1)
    let terminate_result = NtTerminateProcess(pi.hProcess, 0)
    if terminate_result != 0:
        echo "[-] Failed to terminate process"
        quit(1)

    # Close handles
    let close_handle_proc = NtClose(pi.hProcess)
    let close_handle_thread = NtClose(pi.hThread)
    if close_handle_proc != 0 or close_handle_thread != 0:
        echo "[-] Handle closure failed"
        quit(1)
    result = cast[ptr uint8](addr ntdll_buffer[0])


proc replace_ntdll_txt_section*(unhooked_ntdll_txt: pointer, local_ntdll_txt: pointer, local_ntdll_txt_size: uint32) =
    var
        dw_old_protection: ULONG = 0
        current_process = cast[HANDLE](-1)  # UInt64::MAX equivalent
        region_size = local_ntdll_txt_size.SIZE_T
        status: NTSTATUS

    # First protection change to PAGE_EXECUTE_WRITECOPY
    status = NtProtectVirtualMemory(
        current_process,
        addr local_ntdll_txt,
        addr region_size,
        PAGE_EXECUTE_WRITECOPY,
        addr dw_old_protection
    )
    if status != 0:
        echo "[-] Failed to change memory protection"
        quit(1)

    # Perform the memory copy
    let src = cast[ptr UncheckedArray[byte]](unhooked_ntdll_txt)
    let dst = cast[ptr UncheckedArray[byte]](local_ntdll_txt)
    for i in 0..<local_ntdll_txt_size:
        dst[i] = src[i]
    
    # Restore original protection
    status = NtProtectVirtualMemory(
        current_process,
        addr local_ntdll_txt,
        addr region_size,
        dw_old_protection,
        addr dw_old_protection
    )
    if status != 0:
        echo "[-] Failed to restore memory protection"
        quit(1)


proc remap_library*() =
  let unhookedNtdllTxt = get_ntdll_from_debug_proc("C:\\Windows\\System32\\notepad.exe")
  let currentProcess = cast[HANDLE](-1)
  let localNtdllHandle = custom_get_module_address(currentProcess, "ntdll.dll")

  if localNtdllHandle == 0:
      echo "[-] Failed to get ntdll.dll base address"
      quit(1)

  let text_section = get_text_section_info(cast[PVOID](localNtdllHandle))
  let localNtdllTxtBase = text_section[0]
  let localNtdllTxtSize = text_section[1]
  let localNtdllTxt = cast[pointer](localNtdllHandle + localNtdllTxtBase)  # Convert to pointer

  echo fmt"[+] Replacing 0x{localNtdllTxtSize:X} bytes from 0x{cast[uint64](unhookedNtdllTxt):X} to 0x{cast[uint64](localNtdllTxt):X}"
  replace_ntdll_txt_section(cast[pointer](unhookedNtdllTxt), localNtdllTxt, localNtdllTxtSize.uint32)


proc main() =
  var
    jsonFile = "lock.json"
    shouldRemap = false

  for kind, key, val in getopt():
    case kind
    of cmdArgument:
      discard  # Handle positional arguments if needed
    of cmdLongOption, cmdShortOption:
      case key
      of "j", "json":
        if val.len > 0:
          jsonFile = val
      of "r", "remap":
        shouldRemap = true
      else:
        echo "Unknown option: ", key
        quit(1)
    of cmdEnd:
      discard

  if shouldRemap:
    remap_library()
  lock(jsonFile)


when isMainModule:
  main()

File: src/bin/lock.rs
use std::{
    ffi::{c_void, OsStr},
    fs::File,
    io::Write,
    iter::once,
    mem::{size_of, zeroed},
    os::windows::ffi::OsStrExt,
    ptr::null_mut,
};
use clap::Parser;
use winapi::{
    shared::{
        basetsd::SIZE_T,
        minwindef::{DWORD, ULONG},
        ntdef::{HANDLE, NTSTATUS, PVOID},
    },
    um::{
        processthreadsapi::{
            CreateProcessW, PROCESS_INFORMATION, STARTUPINFOW,
        },
        winbase::DEBUG_PROCESS,
        winnt::{OSVERSIONINFOW, RTL_OSVERSIONINFOW},
    },
};


const PROCESS_BASIC_INFORMATION_CLASS: ULONG = 0;
const STATUS_ACCESS_VIOLATION: i32 = 0xC0000005u32 as i32;
const STATUS_INVALID_PARAMETER: i32 = 0x8000000Du32 as i32;
const NT_SUCCESS: fn(NTSTATUS) -> bool = |status| status >= 0;
const STATUS_SUCCESS: NTSTATUS = 0;
const PROCESS_DEBUG_OBJECT_HANDLE: ULONG = 0x1e;
const PAGE_EXECUTE_WRITECOPY: u32 = 0x80;


unsafe extern "system" {
    fn RtlGetVersion(lpVersionInformation: *mut RTL_OSVERSIONINFOW) -> NTSTATUS;
    fn NtReadVirtualMemory(ProcessHandle: HANDLE, BaseAddress: PVOID, Buffer: PVOID, BufferSize: usize, NumberOfBytesRead: *mut usize) -> NTSTATUS;
    fn NtQueryInformationProcess(ProcessHandle: HANDLE, ProcessInformationClass: ULONG, ProcessInformation: PVOID, ProcessInformationLength: ULONG, ReturnLength: *mut ULONG) -> NTSTATUS;
    fn NtRemoveProcessDebug(ProcessHandle: HANDLE, DebugObjectHandle: HANDLE) -> NTSTATUS;
    fn NtTerminateProcess(ProcessHandle: HANDLE, ExitStatus: NTSTATUS) -> NTSTATUS;
    fn NtClose(Handle: HANDLE) -> NTSTATUS;
    fn NtProtectVirtualMemory(ProcessHandle: HANDLE, BaseAddress: *mut *mut c_void, RegionSize: *mut SIZE_T, NewProtect: u32, OldProtect: *mut ULONG,) -> NTSTATUS;
    fn NtWriteVirtualMemory( ProcessHandle: HANDLE, BaseAddress: *mut c_void, Buffer: *const c_void, NumberOfBytesToWrite: usize, NumberOfBytesWritten: *mut usize,) -> NTSTATUS;
}


#[derive(Debug)] pub struct TextSectionInfo { pub base_of_code: DWORD, pub size_of_code: DWORD,}


fn lock(filename: &str) -> Result<(), Box<dyn std::error::Error>> {
    let mut osvi = OSVERSIONINFOW {
        dwOSVersionInfoSize: std::mem::size_of::<OSVERSIONINFOW>() as DWORD,
        dwMajorVersion: 0,
        dwMinorVersion: 0,
        dwBuildNumber: 0,
        dwPlatformId: 0,
        szCSDVersion: [0; 128],
    };

    unsafe {
        if RtlGetVersion(&mut osvi) == 0 {
            let mut file = File::create(filename)?;
            write!(
                file,
                "[{{\"field0\":\"{}\",\"field1\":\"{}\",\"field2\":\"{}\"}}]",
                osvi.dwMajorVersion, osvi.dwMinorVersion, osvi.dwBuildNumber
            )?;
            println!("[+] File {} generated.", filename);
        } else {
            return Err("RtlGetVersion call failed".into());
        }
    }
    Ok(())
}


unsafe fn read_remote_int_ptr(process: HANDLE, address: PVOID) -> Option<PVOID> {
    let mut buffer = [0u8; 8];
    let mut bytes_read = 0usize;
    let status = NtReadVirtualMemory(
        process,
        address,
        buffer.as_mut_ptr() as PVOID,
        buffer.len(),
        &mut bytes_read,
    );

    if status != 0 && status != STATUS_ACCESS_VIOLATION && status != STATUS_INVALID_PARAMETER {
        eprintln!(
            "[-] Error calling NtReadVirtualMemory (ReadRemoteIntPtr). NTSTATUS: 0x{:X} reading address {:p}",
            status, address
        );
        return None;
    }

    let value = *(buffer.as_ptr() as *const usize) as PVOID;
    Some(value)
}


unsafe fn read_remote_wstr(process: HANDLE, address: PVOID) -> String {
    let mut buffer = [0u8; 256];
    let mut bytes_read = 0usize;
    let status = NtReadVirtualMemory(
        process,
        address,
        buffer.as_mut_ptr() as PVOID,
        buffer.len(),
        &mut bytes_read,
    );

    if status != 0 && status != STATUS_ACCESS_VIOLATION && status != STATUS_INVALID_PARAMETER {
        eprintln!(
            "[-] Error calling NtReadVirtualMemory (ReadRemoteWStr). NTSTATUS: 0x{:X} reading address {:p}",
            status, address
        );
    }

    let mut output = Vec::new();
    for i in (0..buffer.len() - 1).step_by(2) {
        let wchar = u16::from_le_bytes([buffer[i], buffer[i + 1]]);
        if wchar == 0 {
            break;
        }
        output.push(wchar);
    }

    String::from_utf16_lossy(&output)
}


unsafe fn get_module_address(dll_name: &str) -> Option<PVOID> {
    let process_basic_information_size = 48;
    let peb_offset = 0x8;
    let ldr_offset = 0x18;
    let in_initialization_order_module_list_offset = 0x30;
    let flink_dllbase_offset = 0x20;
    let flink_buffer_offset = 0x50;

    let mut pbi_buffer = [0u8; 48];
    let mut return_length: ULONG = 0;
    let h_process: HANDLE = -1isize as HANDLE;

    let status = NtQueryInformationProcess(
        h_process,
        PROCESS_BASIC_INFORMATION_CLASS,
        pbi_buffer.as_mut_ptr() as PVOID,
        process_basic_information_size as ULONG,
        &mut return_length,
    );

    if !NT_SUCCESS(status) {
        eprintln!("[-] Error calling NtQueryInformationProcess. NTSTATUS: 0x{:08X}", status);
        return None;
    }

    let peb_address_ptr = pbi_buffer.as_ptr().add(peb_offset) as *const PVOID;
    let peb_address = read_remote_int_ptr(h_process, peb_address_ptr as PVOID)?;

    let ldr_pointer = (peb_address as usize + ldr_offset) as PVOID;
    let ldr_address = read_remote_int_ptr(h_process, ldr_pointer)?;

    let in_initialization_order_module_list = (ldr_address as usize + in_initialization_order_module_list_offset) as PVOID;
    let mut next_flink = read_remote_int_ptr(h_process, in_initialization_order_module_list)?;

    while !next_flink.is_null() {
        next_flink = (next_flink as usize - 0x10) as PVOID;

        let dll_base_ptr = (next_flink as usize + flink_dllbase_offset) as PVOID;
        let dll_base = read_remote_int_ptr(h_process, dll_base_ptr)?;

        let buffer_ptr = (next_flink as usize + flink_buffer_offset) as PVOID;
        let buffer = read_remote_int_ptr(h_process, buffer_ptr)?;
        let base_dll_name = read_remote_wstr(h_process, buffer);

        if base_dll_name.to_lowercase() == dll_name.to_lowercase() {
            return Some(dll_base);
        }

        let next_flink_ptr = (next_flink as usize + 0x10) as PVOID;
        next_flink = read_remote_int_ptr(h_process, next_flink_ptr)?;
    }

    None
}


pub unsafe fn get_text_section_info(ntdll_address: *mut u8) -> Option<TextSectionInfo> {
    let h_process =  -1isize as HANDLE;
    let mut bytes_read = 0;

    // Check MZ Signature (2 bytes)
    let mut signature_dos_header = [0u8; 2];
    if !NT_SUCCESS(NtReadVirtualMemory(
        h_process,
        ntdll_address as _,
        signature_dos_header.as_mut_ptr() as _,
        2,
        &mut bytes_read,
    )) || bytes_read != 2
    {
        eprintln!("[-] Error reading DOS header signature");
        return None;
    }

    if signature_dos_header[0] != b'M' || signature_dos_header[1] != b'Z' {
        eprintln!("[-] Incorrect DOS header signature");
        return None;
    }

    // Read e_lfanew (4 bytes) at offset 0x3C
    let mut e_lfanew = 0u32;
    if !NT_SUCCESS(NtReadVirtualMemory(
        h_process,
        ntdll_address.add(0x3C) as _,
        &mut e_lfanew as *mut _ as _,
        4,
        &mut bytes_read,
    )) || bytes_read != 4
    {
        eprintln!("[-] Error reading e_lfanew");
        return None;
    }

    // Check PE Signature (2 bytes)
    let mut signature_nt_header = [0u8; 2];
    if !NT_SUCCESS(NtReadVirtualMemory(
        h_process,
        ntdll_address.add(e_lfanew as usize) as _,
        signature_nt_header.as_mut_ptr() as _,
        2,
        &mut bytes_read,
    )) || bytes_read != 2
    {
        eprintln!("[-] Error reading NT header signature");
        return None;
    }

    if signature_nt_header[0] != b'P' || signature_nt_header[1] != b'E' {
        eprintln!("[-] Incorrect NT header signature");
        return None;
    }

    // Check Optional Headers Magic field value (2 bytes)
    let mut optional_header_magic = 0u16;
    if !NT_SUCCESS(NtReadVirtualMemory(
        h_process,
        ntdll_address.add(e_lfanew as usize + 24) as _,
        &mut optional_header_magic as *mut _ as _,
        2,
        &mut bytes_read,
    )) || bytes_read != 2
    {
        eprintln!("[-] Error reading Optional Header Magic");
        return None;
    }

    if optional_header_magic != 0x20B && optional_header_magic != 0x10B {
        eprintln!("[-] Incorrect Optional Header Magic field value");
        return None;
    }

    // Read SizeOfCode (4 bytes)
    let mut sizeofcode = 0u32;
    if !NT_SUCCESS(NtReadVirtualMemory(
        h_process,
        ntdll_address.add(e_lfanew as usize + 24 + 4) as _,
        &mut sizeofcode as *mut _ as _,
        4,
        &mut bytes_read,
    )) || bytes_read != 4
    {
        eprintln!("[-] Error reading SizeOfCode");
        return None;
    }

    // Read BaseOfCode (4 bytes)
    let mut baseofcode = 0u32;
    if !NT_SUCCESS(NtReadVirtualMemory(
        h_process,
        ntdll_address.add(e_lfanew as usize + 24 + 20) as _,
        &mut baseofcode as *mut _ as _,
        4,
        &mut bytes_read,
    )) || bytes_read != 4
    {
        eprintln!("[-] Error reading BaseOfCode");
        return None;
    }

    Some(TextSectionInfo {
        base_of_code: baseofcode,
        size_of_code: sizeofcode,
    })
}


pub unsafe fn get_ntdll_from_debug_proc(process_path: &str) -> Vec<u8> {
    let mut si: STARTUPINFOW = zeroed();
    si.cb = size_of::<STARTUPINFOW>() as DWORD;
    let mut pi: PROCESS_INFORMATION = zeroed();
    let mut debug_object_handle: HANDLE = null_mut();
    let mut return_length: ULONG = 0;

    // Convert path to wide string
    let wide_path: Vec<u16> = OsStr::new(process_path).encode_wide().chain(once(0)).collect();

    // Create suspended process with debug flag
    let success = CreateProcessW(
        wide_path.as_ptr(),
        null_mut(),
        null_mut(),
        null_mut(),
        0,
        DEBUG_PROCESS,
        null_mut(),
        null_mut(),
        &mut si,
        &mut pi,
    );

    if success == 0 {
        eprintln!("[-] CreateProcess failed");
        std::process::exit(1);
    }

    // Get local ntdll address and .text section info
    let local_ntdll_handle = match get_module_address("ntdll.dll") {
        Some(handle) => handle,
        None => {
            eprintln!("[-] Failed to locate ntdll.dll in current process");
            std::process::exit(1);
        }
    };

    let text_info = match get_text_section_info(local_ntdll_handle as *mut u8) {
        Some(info) => info,
        None => {
            eprintln!("[-] Failed to parse local ntdll");
            std::process::exit(1);
        }
    };

    let local_ntdll_text = (local_ntdll_handle as usize + text_info.base_of_code as usize) as PVOID;
    let text_size = text_info.size_of_code as usize;
    let mut ntdll_buffer = vec![0u8; text_size];
    let mut bytes_read: usize = 0;

    let status = NtReadVirtualMemory(
        pi.hProcess,
        local_ntdll_text,
        ntdll_buffer.as_mut_ptr() as PVOID,
        text_size,
        &mut bytes_read,
    );

    if !NT_SUCCESS(status) {
        eprintln!("[-] Read operation failed");
        std::process::exit(1);
    }

    // Query debug object handle
    let status = NtQueryInformationProcess(
        pi.hProcess,
        PROCESS_DEBUG_OBJECT_HANDLE,
        &mut debug_object_handle as *mut _ as PVOID,
        size_of::<HANDLE>() as ULONG,
        &mut return_length,
    );

    if !NT_SUCCESS(status) {
        eprintln!("[-] Failed to get debug object handle");
        std::process::exit(1);
    }

    // Cleanup
    let status = NtRemoveProcessDebug(pi.hProcess, debug_object_handle);
    if !NT_SUCCESS(status) {
        eprintln!("[-] Failed to remove process debug");
        std::process::exit(1);
    }

    let status = NtTerminateProcess(pi.hProcess, 0);
    if !NT_SUCCESS(status) {
        eprintln!("[-] Failed to terminate process");
        std::process::exit(1);
    }

    if !NT_SUCCESS(NtClose(pi.hProcess)) || !NT_SUCCESS(NtClose(pi.hThread)) {
        eprintln!("[-] Handle closure failed");
        std::process::exit(1);
    }

    ntdll_buffer
}


pub unsafe fn replace_ntdll_txt_section(
    unhooked_ntdll_txt: *const u8,
    local_ntdll_txt: *mut u8,
    local_ntdll_txt_size: u32,
) {
    let current_process = -1isize as HANDLE;
    let mut region_size: SIZE_T = local_ntdll_txt_size as SIZE_T;
    let mut base_address: *mut c_void = local_ntdll_txt as *mut c_void;
    let mut old_protection: ULONG = 0;


    // Change protection to PAGE_EXECUTE_WRITECOPY
    let status = NtProtectVirtualMemory(
        current_process,
        &mut base_address,
        &mut region_size,
        PAGE_EXECUTE_WRITECOPY,
        &mut old_protection,
    );
    if status != STATUS_SUCCESS {
        eprintln!("[-] Failed to change memory protection");
        std::process::exit(1);
    }

    // Copy memory from unhooked to hooked
    let mut bytes_written: usize = 0;
    let status = NtWriteVirtualMemory(
        current_process,
        local_ntdll_txt as *mut c_void,
        unhooked_ntdll_txt as *const c_void,
        local_ntdll_txt_size as usize,
        &mut bytes_written,
    );
    if status != STATUS_SUCCESS || bytes_written != local_ntdll_txt_size as usize {
        eprintln!(
            "[-] Failed to write memory: status={:#X}, written={}",
            status, bytes_written
        );
        std::process::exit(1);
    }

    // Restore original protection
    let mut region_size: SIZE_T = local_ntdll_txt_size as SIZE_T;
    let mut base_address: *mut c_void = local_ntdll_txt as *mut c_void;
    let status = NtProtectVirtualMemory(
        current_process,
        &mut base_address,
        &mut region_size,
        old_protection,
        &mut old_protection,
    );
    if status != STATUS_SUCCESS {
        eprintln!("[-] Failed to restore memory protection");
        std::process::exit(1);
    }
}


pub unsafe fn remap_library(){    
    let unhooked_ntdll_txt = get_ntdll_from_debug_proc(r"C:\Windows\System32\notepad.exe");
    
    // Get local ntdll.dll base address
    let local_ntdll_handle = match get_module_address("ntdll.dll") {
        Some(handle) => handle,
        None => {
            eprintln!("[-] Failed to get ntdll.dll base address");
            std::process::exit(1);
        }
    };

    // Get text section information
    let text_section = match get_text_section_info(local_ntdll_handle as *mut u8) {
        Some(info) => info,
        None => {
            eprintln!("[-] Failed to get text section info");
            std::process::exit(1);
        }
    };

    let local_ntdll_txt_base = text_section.base_of_code;
    let local_ntdll_txt_size = text_section.size_of_code;
    let local_ntdll_txt = (local_ntdll_handle as usize + local_ntdll_txt_base as usize) as *mut u8;
    let unhooked_ntdll_txt_ptr = unhooked_ntdll_txt.as_ptr();
    
    println!("[+] Replacing 0x{:X} bytes from 0x{:X} to 0x{:X}", local_ntdll_txt_size, unhooked_ntdll_txt_ptr as usize, local_ntdll_txt as usize);
    replace_ntdll_txt_section(unhooked_ntdll_txt_ptr, local_ntdll_txt, local_ntdll_txt_size);
}


#[derive(Parser)]
#[command(author, version, about, long_about = None)]
struct Args {
    #[arg(short = 'r', long)]
    remap: bool,

    #[arg(short = 'j', long, default_value = "lock.json")]
    json_file: String,
}


fn main() {
    let args = Args::parse();
    unsafe {
        if args.remap {
            remap_library();
        }
        if let Err(e) = lock(&args.json_file) {
            eprintln!("Error: {}", e);
        }
    }
}